src/wayland/shell/wlr_layer/mod.rs

```diff
@@ -0,0 +1,454 @@
+//! Utilities for handling shell surfaces with the `wlr_layer_shell` protocol
+//!
+//! This interface should be suitable for the implementation of many desktop shell components,
+//! and a broad number of other applications that interact with the desktop.
+//!
+//! ### Initialization
+//!
+//! To initialize this handler, simple use the [`wlr_layer_shell_init`] function provided in this module.
+//! You need to provide a closure that will be invoked whenever some action is required from you,
+//! are represented by the [`LayerShellRequest`] enum.
+//!
+//! ```no_run
+//! # extern crate wayland_server;
+//! #
+//! use smithay::wayland::shell::wlr_layer::{wlr_layer_shell_init, LayerShellRequest};
+//!
+//! # let mut display = wayland_server::Display::new();
+//! let (shell_state, _) = wlr_layer_shell_init(
+//!     &mut display,
+//!     // your implementation
+//!     |event: LayerShellRequest, dispatch_data| { /* handle the shell requests here */ },
+//!     None  // put a logger if you want
+//! );
+//!
+//! // You're now ready to go!
+//! ```
+
+use std::{
+    cell::RefCell,
+    rc::Rc,
+    sync::{Arc, Mutex},
+};
+
+use wayland_protocols::wlr::unstable::layer_shell::v1::server::{zwlr_layer_shell_v1, zwlr_layer_surface_v1};
+use wayland_server::{
+    protocol::{wl_output::WlOutput, wl_surface},
+    DispatchData, Display, Filter, Global, Main,
+};
+
+use crate::{
+    utils::{DeadResource, Logical, Size},
+    wayland::{
+        compositor::{self, Cacheable},
+        Serial, SERIAL_COUNTER,
+    },
+};
+
+mod handlers;
+mod types;
+
+pub use types::{Anchor, ExclusiveZone, KeyboardInteractivity, Layer, Margins};
+
+/// The role of a wlr_layer_shell_surface
+pub static LAYER_SURFACE_ROLE: &str = "zwlr_layer_surface_v1";
```

Comment:
Probably a little better to make this `pub const` here. Static associates a memory address with the value and using `const` allows the values to be inlined a little more easily by rustc.

Reply:
Good point, it's better to use compile time constants for this